src/Logic.hs

```diff
@@ -240,23 +244,38 @@ runBaseAction config = \case
       case shaOrFailed of
         Left failure -> pure $ cont $ Left $ IntegrationFailure (Git.toBaseBranch targetBranch) failure
         Right integratedSha -> pure $ cont $ Right integratedSha
-    TryPromote prBranch sha cont -> do
-      doGit $ ensureCloned config
-      forcePushResult <- doGit $ Git.forcePush sha prBranch
+    TryPromote prBranch sha cont -> doGit $ do
+      ensureCloned config
+      forcePushResult <- Git.forcePush sha prBranch
       case forcePushResult of
         PushRejected _ -> pure $ cont forcePushResult
         PushOk -> do
-          pushResult <- doGit $ Git.push sha (Git.Branch $ Config.branch config)
+          -- TODO: Find a safer way to make sure Github doesn't get confused
+          -- by 2 pushes close together, the delay is all arbitrary and not nice.
+          -- See https://github.com/channable/hoff/issues/196
+          _ <- pure $ threadDelay pushDelayMicroseconds
+          pushResult <- Git.push sha (Git.Branch $ Config.branch config)
           pure $ cont pushResult
 
-    TryPromoteWithTag prBranch sha newTagName newTagMessage cont -> doGit $
-      ensureCloned config >>
-      Git.forcePush sha prBranch >>
-      Git.tag sha newTagName newTagMessage >>=
-      \case TagFailed _ -> cont . (Left "Please check the logs",) <$> Git.push sha (Git.Branch $ Config.branch config)
-            TagOk tagName -> cont . (Right tagName,)
-              <$> Git.pushAtomic [AsRefSpec tagName, AsRefSpec (sha, Git.Branch $ Config.branch config)]
-              <*  Git.deleteTag tagName
+    TryPromoteWithTag prBranch sha newTagName newTagMessage cont -> doGit $ do
+      ensureCloned config
+      forcePushResult <- Git.forcePush sha prBranch
+      case forcePushResult of
+        PushRejected err -> pure $ cont (Left err, forcePushResult)
+        PushOk -> do
+          tagResult <- Git.tag sha newTagName newTagMessage
+          -- TODO: Find a safer way to make sure Github doesn't get confused
+          -- by 2 pushes close together, the delay is all arbitrary and not nice.
+          -- See https://github.com/channable/hoff/issues/196
+          _ <- pure $ threadDelay pushDelayMicroseconds
```

Comment:
I would still move the `threadDelay` up closer to `PushOk` and `Git.tag` nearer to `case tagResult`.

```suggestion
          -- TODO: Find a safer way to make sure Github doesn't get confused
          -- by 2 pushes close together, the delay is all arbitrary and not nice.
          -- See https://github.com/channable/hoff/issues/196
          _ <- pure $ threadDelay pushDelayMicroseconds
          tagResult <- Git.tag sha newTagName newTagMessage
```